### ISSUE TITLE
need a way to delete consumer group

### ISSUE THREAD

**Issue by wz2cool** [apache/rocketmq#3660]
version 4.8.0

For some business reasons, we need create consumer group by ip address dynamically.  but we also need to delete consumer group if the application is down.  

It seems that it doesn't work, I have deleted all consumer groups in rocketmq-console, but I still see them in topic.  I find related issue about it #1416 #1799 #3094 #2558

**Comment by wz2cool** [comment]
find another issue about it #3094   need help

**Comment by panzhi33** [comment]
somebody can fix it,you can see if it is what you want
https://github.com/apache/rocketmq-dashboard/pull/45

**Comment by wz2cool** [comment]
> somebody can fix it,you can see if it is what you want [apache/rocketmq-dashboard#45](https://github.com/apache/rocketmq-dashboard/pull/45)

I'm afraid that it doesn't work . it seems that just delete offset, not delete relation of topic

**Comment by wz2cool** [comment]
here is code, I use
```
private void deleteSubscriptionGroup(String topic, String consumerGroup)
            throws RemotingException, MQClientException, InterruptedException, MQBrokerException {
        final TopicRouteData topicRouteData = this.mqAdmin.examineTopicRouteInfo(topic);
        if (CollectionUtils.isEmpty(topicRouteData.getBrokerDatas())) {
            return;
        }

        for (BrokerData brokerData : topicRouteData.getBrokerDatas()) {
            for (Map.Entry<Long, String> entry : brokerData.getBrokerAddrs().entrySet()) {
                String brokerAddr = entry.getValue();
                this.mqAdmin.deleteSubscriptionGroup(brokerAddr, consumerGroup, true);
            }
        }
    }
```

**Comment by panzhi33** [comment]
> > somebody can fix it,you can see if it is what you want [apache/rocketmq-dashboard#45](https://github.com/apache/rocketmq-dashboard/pull/45)
> 
> I'm afraid that it doesn't work . it seems that just delete offset, not delete relation of topic

The subscription relationship is maintained by the client，just delete the consumer offset

**Comment by wz2cool** [comment]
> > > somebody can fix it,you can see if it is what you want [apache/rocketmq-dashboard#45](https://github.com/apache/rocketmq-dashboard/pull/45)
> > 
> > 
> > I'm afraid that it doesn't work . it seems that just delete offset, not delete relation of topic
> 
> The subscription relationship is maintained by the client，just delete the consumer offset

I think something wrong in server side.

![image](https://user-images.githubusercontent.com/11556152/146531123-65cde306-a312-4c6c-b9ab-eab605fe1832.png)

**Comment by wz2cool** [comment]
i write a example to filter valid consumer group

```
 private Set<String> listValidConsumerGroups(String topic)
            throws RemotingException, MQClientException, InterruptedException, MQBrokerException {
        final TopicRouteData topicRouteData = this.mqAdmin.examineTopicRouteInfo(topic);
        if (CollectionUtils.isEmpty(topicRouteData.getBrokerDatas())) {
            return new HashSet<>();
        }
        Set<String> validGroupSet = new HashSet<>();
        for (BrokerData brokerData : topicRouteData.getBrokerDatas()) {
            for (Map.Entry<Long, String> entry : brokerData.getBrokerAddrs().entrySet()) {
                String brokerAddr = entry.getValue();
                long timeoutMillis = TimeUnit.SECONDS.toMillis(1);
                final SubscriptionGroupWrapper allSubscriptionGroup = this.mqAdmin.getAllSubscriptionGroup(brokerAddr, timeoutMillis);
                ConcurrentMap<String, SubscriptionGroupConfig> subscriptionGroupTable = allSubscriptionGroup.getSubscriptionGroupTable();
                validGroupSet.addAll(subscriptionGroupTable.keySet());
            }
        }
        final GroupList consumeGroupList = this.mqAdmin.queryTopicConsumeByWho(topic);
        // 删除消费组是无法删除topic 和消费组的关系的， 所以我们要和现在的组做一次交集运算
        final HashSet<String> relationGroups = consumeGroupList.getGroupList();
        validGroupSet.retainAll(relationGroups);
        return validGroupSet;
    }
```

![image](https://user-images.githubusercontent.com/11556152/146541023-25520de2-f816-47df-bd7c-38c9e518f4e4.png)

**Comment by panzhi33** [comment]
Need to upgrade to version 4.9 or above。
org.apache.rocketmq.broker.offset.ConsumerOffsetManager#removeOffset  
Version 4.8 does not have this method, and consumerOffset cannot be deleted

**Comment by wz2cool** [comment]
> Need to upgrade to version 4.9 or above。 org.apache.rocketmq.broker.offset.ConsumerOffsetManager#removeOffset Version 4.8 does not have this method, and consumerOffset cannot be deleted

thank you for your response.
yes , i indeed test it by using 4.9.2, but queryTopicConsumeByWho contains deleted customer group

![image](https://user-images.githubusercontent.com/11556152/146730580-9b42b14f-0be9-4bbf-954b-03cedc668ea4.png)

**Comment by wz2cool** [comment]
@panzhi33  I write a demo to show this problem

[rocketmq-consumer.zip](https://github.com/apache/rocketmq/files/7743392/rocketmq-consumer.zip)

**Comment by wz2cool** [comment]
Read the sourcecode of rocketmq-dashboard. it also call queryTopicConsumeByWho(topic) . it's why we can see deleted customer group in UI

![image](https://user-images.githubusercontent.com/11556152/146732344-c3791fcb-6537-40bf-bc0d-8d4f16fbe858.png)

**Comment by panzhi33** [comment]
queryTopicConsumeByWho  This method checks the subscription relationship. If you want to delete the group, the consumer will stop.
After reading your demo, you deleted the consumer without stopping it. This is a non-standard operation. If you want to delete a group, it must be that the group is no longer in use.

**Comment by wz2cool** [comment]
> queryTopicConsumeByWho This method checks the subscription relationship. If you want to delete the group, the consumer will stop. After reading your demo, you deleted the consumer without stopping it. This is a non-standard operation. If you want to delete a group, it must be that the group is no longer in use.

yes, you are right , it works in rocketmq(4.9.2) will test in rocketmq(4.8.0) ,  thank you for your help

**Comment by wz2cool** [comment]
> queryTopicConsumeByWho This method checks the subscription relationship. If you want to delete the group, the consumer will stop. After reading your demo, you deleted the consumer without stopping it. This is a non-standard operation. If you want to delete a group, it must be that the group is no longer in use.

yes you are right , your response make me clear ("no longer in use "),  Everything seems ok  if call DefaultMQPushConsumer#shutdown method.

But, if close the application without call DefaultMQPushConsumer#shutdown, it will keep the relation of topic

**Comment by wz2cool** [comment]
after some tests , I have confirmed this issue has been resolved in rocketmq 4.9 .  
this issue will happen when using rocketmq 4.8 even if using rocketmq-tool 4.9.2

@panzhi33 I found some related issues opened , could you  help to close it?

**Comment by panzhi33** [comment]
@duhenglucky
